Optimize IOperation.Children Implementation

333fred opened the pull request:
Optimizes the implemention of IOperation.Children by using a struct enumerable/enumerator internally, with slot/index tracking, and exposing this as ChildOperations on Operation internally. Uses of IOperation.Children in internal compiler code such as Descendents() and OperationWalker have been updated to use this new property, making significant improvements to memory usage. I've taken some benchmarks with 3 different scenarios:

1. master (2766389028b)
2. old-children. This uses the same Children implementation as master, but otherwise uses the new non-lazy generation in features/ioperation.
3. better-children. This is the implementation of Children in this branch.

### AnalyzerRunner Results

Note: times are somewhat variable. I've run the benchmarks a few times and while on the whole ms taken does appear to be improved, I can't say it has with statistic certainty. Memory usage, on the other hand, is much more deterministic, and I can confidently say that we save a few GBs of allocations. This run was performed using Microsoft.CodeAnalysis.Analyzers (which is heavily IOperation-based).

**branch**|** time**|** bytes allocated**
:-----:|:-----:|:-----:
master (2766389028b)| 536240ms|123,258,378,656
old-children| 538872ms|122,012,083,264
better-children| 526784ms|119,468,190,816

### dotnet/performance Roslyn Benchmarks

The main benchmark to pay attention to here is GetDiagnosticsWithAnalyzers, but I included the rest for completeness. There's not a clear-cut win in terms of execution times, but there is nearly 100MB saved in memory usage in our bench.

#### master

|                      Method |        Job | UnrollFactor |        Mean |     Error |    StdDev |      Median |         Min |         Max |       Gen 0 |      Gen 1 |     Gen 2 | Allocated |
|---------------------------- |----------- |------------- |------------:|----------:|----------:|------------:|------------:|------------:|------------:|-----------:|----------:|----------:|
|       CompileMethodsAndEmit | Job-ZZBXBK |           16 |  5,243.7 ms | 103.96 ms | 289.79 ms |  5,175.5 ms |  4,906.4 ms |  6,126.8 ms |  82000.0000 | 23000.0000 |         - |    497 MB |
|           SerializeMetadata | Job-ZZBXBK |           16 |    320.2 ms |   3.17 ms |   5.64 ms |    319.3 ms |    308.5 ms |    334.6 ms |   4000.0000 |  1000.0000 |         - |     34 MB |
|              GetDiagnostics | Job-KZXNEV |            1 |  4,314.1 ms |  97.91 ms | 287.16 ms |  4,209.7 ms |  3,901.5 ms |  5,160.1 ms |  73000.0000 | 21000.0000 |         - |    379 MB |
| GetDiagnosticsWithAnalyzers | Job-KZXNEV |            1 | 11,486.9 ms | 112.89 ms | 120.79 ms | 11,496.5 ms | 11,295.6 ms | 11,695.0 ms | 202000.0000 | 59000.0000 | 1000.0000 |  1,215 MB |

#### old-children

|                      Method |        Job | UnrollFactor |        Mean |    Error |    StdDev |      Median |         Min |         Max |       Gen 0 |      Gen 1 |     Gen 2 | Allocated |
|---------------------------- |----------- |------------- |------------:|---------:|----------:|------------:|------------:|------------:|------------:|-----------:|----------:|----------:|
|       CompileMethodsAndEmit | Job-XXOYTR |           16 |  5,047.9 ms | 34.00 ms |  31.81 ms |  5,040.0 ms |  4,992.4 ms |  5,097.3 ms |  82000.0000 | 24000.0000 |         - |    497 MB |
|           SerializeMetadata | Job-XXOYTR |           16 |    325.8 ms |  2.97 ms |   5.13 ms |    324.6 ms |    317.4 ms |    338.2 ms |   3000.0000 |  1000.0000 |         - |     34 MB |
|              GetDiagnostics | Job-QCCYOV |            1 |  3,708.4 ms | 29.22 ms |  24.40 ms |  3,704.9 ms |  3,669.2 ms |  3,765.8 ms |  73000.0000 | 17000.0000 |         - |    378 MB |
| GetDiagnosticsWithAnalyzers | Job-QCCYOV |            1 | 11,112.4 ms | 99.31 ms | 135.93 ms | 11,096.5 ms | 10,884.7 ms | 11,436.8 ms | 195000.0000 | 55000.0000 | 1000.0000 |  1,178 MB |

#### better-children

|                      Method |        Job | UnrollFactor |        Mean |     Error |    StdDev |      Median |         Min |         Max |       Gen 0 |      Gen 1 |     Gen 2 | Allocated |
|---------------------------- |----------- |------------- |------------:|----------:|----------:|------------:|------------:|------------:|------------:|-----------:|----------:|----------:|
|       CompileMethodsAndEmit | Job-UBJCRB |           16 |  5,113.7 ms |  40.92 ms |  34.17 ms |  5,097.6 ms |  5,078.3 ms |  5,198.1 ms |  81000.0000 | 22000.0000 |         - |    497 MB |
|           SerializeMetadata | Job-UBJCRB |           16 |    336.1 ms |   3.67 ms |  10.71 ms |    335.6 ms |    315.6 ms |    365.7 ms |   4000.0000 |  1000.0000 |         - |     34 MB |
|              GetDiagnostics | Job-OIOLXP |            1 |  3,768.2 ms |  23.62 ms |  20.94 ms |  3,765.4 ms |  3,741.8 ms |  3,821.4 ms |  72000.0000 | 19000.0000 |         - |    378 MB |
| GetDiagnosticsWithAnalyzers | Job-OIOLXP |            1 | 11,328.5 ms | 113.10 ms | 121.01 ms | 11,317.0 ms | 11,155.9 ms | 11,573.5 ms | 193000.0000 | 60000.0000 | 2000.0000 |  1,156 MB |

### RPS

A test insertion with this payload is [here](https://devdiv.visualstudio.com/DevDiv/_git/VS/pullrequest/291438). The DDRIT failures are unrelated, and a fix is being merged into dotnet/roslyn shortly.

333fred commented:
Commit 1 is the only change that needs to be brought over from https://github.com/dotnet/roslyn/pull/49534. The issue itself exists in master, but is verified in this branch because of the additional Debug.Asserts, and was then triggered by AnalyzerRunner.

Commit 2 is the actual implementation. I used a struct enumerator with a slot-based approach: every `IOperation` or `ImmutableArray<IOperation>` has a slot. I also track the current index for the arrays as appropriate. Iterating advances through the slots and indexes. I've also updated our extension methods and walkers to use the new method, rather than `Children`, leading to the memory improvements detailed above.

333fred commented:
@dotnet/roslyn-compiler @AlekseyTs @mavasani @sharwell please take a look.

sharwell commented on src/Compilers/Core/Portable/Operations/Operation.Enumerable.cs:
```diff
@@ -0,0 +1,102 @@
+            IEnumerator IEnumerable.GetEnumerator() => this.GetEnumerator();
+        }
+
+        internal struct Enumerator : IEnumerator<IOperation>
```
📝 IIRC, you can remove the explicit implementation of `IEnumerator<IOperation>` to ensure this type does not get boxed during usage

333fred replied:
I don't see how I could. It still needs to implement the interface, and if it's being used as `IEnumerator`, it's already been boxed anyway.

sharwell commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6237,14 +6237,14 @@ internal override IOperation VisitNoneOperation(IOperation operation, int? captu
         {
             Debug.Assert(_currentStatement == operation);
-            VisitStatements(operation.Children.ToImmutableArray());
+            VisitStatements(((Operation)operation).ChildOperations.ToImmutableArray());
```
❔ Is there any value in removing these calls to `ToImmutableArray`?

333fred replied:
Well, it removes at least one boxing, and in some cases an array allocation. More generally, I removed all calls to `Children` in the compiler product layer (not the test layers).

sharwell replied:
I mean is it possible/valuable to try and change this to:

```suggestion
            VisitStatements(((Operation)operation).ChildOperations);
```

333fred replied:
I don't believe so. This helper is used for a number of cases, and not all of them are visiting all children of a node.

333fred commented:
@dotnet/roslyn-compiler @AlekseyTs for reviews please.

AlekseyTs commented on src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.Verifier.cs:
```diff
@@ -51,6 +51,15 @@ private bool ModelHasErrors(Tree tree)
+                {
+                    if (IsImmutableArray(prop.Type, out _) && prop.Type.Contains("?"))
+                    {
+                        Console.WriteLine($"{abstractNode.Name}.{prop.Name} has nullable IOperation elements. This is not allowed in IOperation and will mess up Children generation.");
```
>Console.WriteLine($"{abstractNode.Name}.{prop.Name} has nullable IOperation elements. This is not allowed in IOperation and will mess up Children generation."); [](start = 24, length = 160)

Why not handle the nulls?

333fred replied:
Simplicity. We don't have any of these today, and not having to do extra null checks makes generating the code a bit simpler. I'd have to be convinced that exposing an array of nullable operations in the first place would be a good idea, but this will ensure that if we ever do have a reason to do so, we don't forget to update the implementation of the enumerator methods.

---
In reply to: [539571031](https://github.com/dotnet/roslyn/pull/49804#discussion_r539571031) [](ancestors = 539571031)

AlekseyTs commented:
>                 0 when index < Operations.Length => Operations[index],

This is really unreadable, IMHO. Consider either moving the ```=> ...``` part on a new line, or using a regular switch statement. I believe that dense code is bad for readability. 

---
Refers to: src/Compilers/Core/Portable/Generated/Operations.Generated.cs:3335 in 10fdb5c. [](commit_id = 10fdb5c272ab20eb527dc04997ce94518bb4ad35, deletion_comment = False)

AlekseyTs commented:
>                     return (false, int.MinValue, int.MinValue);

It looks like an unexpected ```previousSlot``` value is treated as the end of sequence. Should we be able to distinguish the two and, perhaps, throw on an invalid input.  

---
Refers to: src/Compilers/Core/Portable/Generated/Operations.Generated.cs:3348 in 10fdb5c. [](commit_id = 10fdb5c272ab20eb527dc04997ce94518bb4ad35, deletion_comment = False)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/Operation.Enumerable.cs:
```diff
@@ -0,0 +1,102 @@
+{
+    internal abstract partial class Operation : IOperation
+    {
+        internal readonly struct Enumerable : IEnumerable<IOperation>
```
>Enumerable [](start = 33, length = 10)

Consider adding a doc comment.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/Operation.Enumerable.cs:
```diff
@@ -0,0 +1,102 @@
+    {
+        internal readonly struct Enumerable : IEnumerable<IOperation>
+        {
+            private readonly Operation? _operation;
```
>Operation? [](start = 29, length = 10)

Why nullable?

333fred replied:
The idea was to ensure that we handle `default` instances. I could remove it since we don't have any cases where it's needed.

---
In reply to: [539583374](https://github.com/dotnet/roslyn/pull/49804#discussion_r539583374) [](ancestors = 539583374)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/Operation.Enumerable.cs:
```diff
@@ -0,0 +1,102 @@
+            private int _currentSlot;
+            private int _currentIndex;
+
+            public Enumerator(Operation? operation)
```
>Operation? [](start = 30, length = 10)

It is not obvious why do we need to allow nullable input.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationNodes.cs:
```diff
@@ -26,7 +23,23 @@ internal sealed class NoneOperation : Operation
+        protected override IOperation GetCurrent(int slot, int index)
+            => slot switch
+            {
+                0 when index < Children.Length => Children[index],
```
>0 when index < Children.Length => Children[index], [](start = 16, length = 50)

IMHO. Consider either moving the ```=> ...``` part on a new line, or using a regular switch statement.

AlekseyTs replied:
Similar suggestion applies to other similar changes in this file.

---
In reply to: [539594643](https://github.com/dotnet/roslyn/pull/49804#discussion_r539594643) [](ancestors = 539594643)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationNodes.cs:
```diff
@@ -300,22 +368,55 @@ internal sealed partial class ForToLoopOperation
     {
-        public override IEnumerable<IOperation> Children
+        protected override IOperation GetCurrent(int slot, int index)
+            => (slot, ConditionIsTop) switch
```
> => (slot, ConditionIsTop) switch [](start = 11, length = 33)

I think the logic will be much easier to follow if we separated two cases based on ConditionIsTop rather than trying to "jam" them togather.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationWalker.cs:
```diff
@@ -14,19 +12,14 @@ public abstract class OperationWalker : OperationVisitor
         private int _recursionDepth;
 
-        internal void VisitArray<T>(IEnumerable<T> operations) where T : IOperation
+        internal void VisitArray(IOperation operation)
```
>VisitArray [](start = 22, length = 10)

The name is confusing. "VisitChildOperations"?

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationWalker.cs:
```diff
@@ -14,19 +12,14 @@ public abstract class OperationWalker : OperationVisitor
+        internal void VisitArray(IOperation operation)
         {
-            foreach (var operation in operations)
+            foreach (var child in ((Operation)operation).ChildOperations)
```
>((Operation)operation) [](start = 34, length = 22)

Is it really safe to assume that this public type always deals with nodes based on ```Operation``` type?

333fred replied:
Yes. `IOperation` is marked with our `InternalImplementationOnly` attribute, which will ensure our analyzers (that consumers get automatically when they reference our libraries) mark consumer implementation of these interfaces as errors. We also document that they are reserved for internal implementation only on every interface. Much like our symbol interfaces, if a consumer implements one themselves, there is no guarantee it will work with our APIs.

---
In reply to: [539664626](https://github.com/dotnet/roslyn/pull/49804#discussion_r539664626) [](ancestors = 539664626)

AlekseyTs replied:
> Yes. IOperation is marked with our InternalImplementationOnly attribute, which will ensure our analyzers (that consumers get automatically when they reference our libraries) mark consumer implementation of these interfaces as errors. We also document that they are reserved for internal implementation only on every interface. Much like our symbol interfaces, if a consumer implements one themselves, there is no guarantee it will work with our APIs.

We say pretty much the same for symbols, yet IDE turns around and implements their own. This looks like a breaking change to me. Consider adding this as an optimization when possible, rather than a complete replacement. 


---
In reply to: [539715866](https://github.com/dotnet/roslyn/pull/49804#discussion_r539715866) [](ancestors = 539715866,539664626)

333fred replied:
> Consider adding this as an optimization when possible, rather than a complete replacement. 

I did consider it, actually, but the branches did have a measurable impact on performance.

---
In reply to: [539718172](https://github.com/dotnet/roslyn/pull/49804#discussion_r539718172) [](ancestors = 539718172,539715866,539664626)

AlekseyTs replied:
> I did consider it, actually, but the branches did have a measurable impact on performance.

Then I think this should be approved by Compatibility Council.  

---
In reply to: [539726793](https://github.com/dotnet/roslyn/pull/49804#discussion_r539726793) [](ancestors = 539726793,539718172,539715866,539664626)

333fred replied:
Email setn.

---
In reply to: [539728958](https://github.com/dotnet/roslyn/pull/49804#discussion_r539728958) [](ancestors = 539728958,539726793,539718172,539715866,539664626)

AlekseyTs replied:
> Email sent.

Thanks, we can follow up later based on the decision, if needed. 

---
In reply to: [540550184](https://github.com/dotnet/roslyn/pull/49804#discussion_r540550184) [](ancestors = 540550184,539728958,539726793,539718172,539715866,539664626)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationWalker.cs:
```diff
@@ -63,19 +56,14 @@ public abstract class OperationWalker<TArgument> : OperationVisitor<TArgument, o
+        internal void VisitChildren(IOperation operation, TArgument argument)
         {
-            foreach (var operation in operations)
+            foreach (var child in ((Operation)operation).ChildOperations)
```
>((Operation)operation) [](start = 34, length = 22)

Same question.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -612,7 +612,8 @@ private BoundExpression UnwrapCollectionExpressionIfNullable(BoundExpression col
                         receiverOpt: collectionExpr,
-                        method: nullableValueGetter);
+                        method: nullableValueGetter,
+                        binder: this);
```
>binder: this); [](start = 24, length = 14)

I do not think this change belongs in this PR

333fred replied:
Separated out into [https://github.com/dotnet/roslyn/pull/49885](https://github.com/dotnet/roslyn/pull/49885).

---
In reply to: [539666484](https://github.com/dotnet/roslyn/pull/49804#discussion_r539666484) [](ancestors = 539666484)

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/Constructors.cs:
```diff
@@ -193,7 +197,7 @@ public static BoundCall Synthesized(SyntaxNode syntax, BoundExpression? receiver
                     resultKind: LookupResultKind.Viable,
                     originalMethodsOpt: default,
-                    binderOpt: null,
+                    binder,
```
>binder, [](start = 20, length = 7)

I do not think changes in this file belong in this PR

AlekseyTs commented on src/Compilers/CSharp/Portable/Compiler/MethodBodySynthesizer.cs:
```diff
@@ -540,7 +545,8 @@ internal static BoundBlock ConstructDestructorBody(MethodSymbol method, BoundBlo
                         { WasCompilerGenerated = true },
-                        baseTypeFinalize))
+                        baseTypeFinalize,
+                        binder: null))
```
>binder: null [](start = 24, length = 12)

I do not think changes in this file belong in this PR

AlekseyTs commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IForEachLoopStatement.cs:
```diff
@@ -5568,6 +5568,152 @@ await foreach (string value in pets)
         }
 
+        [Fact]
+        public void AsyncForEach_TestConstantNullableImplementingIEnumerable()
```
>AsyncForEach_TestConstantNullableImplementingIEnumerable [](start = 20, length = 56)

Is this test verifying a fix made in this PR? What in particular?

AlekseyTs commented:
I think changes around the ```binder``` do not belong in this PR. Please separate them into separate PR with appropriate title and description about the reason for the change, etc.

AlekseyTs commented:
Done with review pass (commit 3)

333fred commented:
>                     return (false, int.MinValue, int.MinValue);

> Should we be able to distinguish the two and, perhaps, throw on an invalid input

We could. I didn't think that this would particularly help us catch any bugs, but if you think it's valuable I can take that approach.

---
In reply to: [741991994](https://github.com/dotnet/roslyn/pull/49804#issuecomment-741991994) [](ancestors = 741991994)

---
Refers to: src/Compilers/Core/Portable/Generated/Operations.Generated.cs:3348 in 10fdb5c. [](commit_id = 10fdb5c272ab20eb527dc04997ce94518bb4ad35, deletion_comment = False)

AlekseyTs commented:
>                     return (false, int.MinValue, int.MinValue);

> We could. I didn't think that this would particularly help us catch any bugs, but if you think it's valuable I can take that approach.

Explicitly transitioning to a dedicated "eof" slot (#1 in this case) and staying there might be clearer than to jumping to default, especially with ```goto default```. And the code generator could be somewhat simplified. 


---
In reply to: [742151554](https://github.com/dotnet/roslyn/pull/49804#issuecomment-742151554) [](ancestors = 742151554,741991994)

---
Refers to: src/Compilers/Core/Portable/Generated/Operations.Generated.cs:3348 in 10fdb5c. [](commit_id = 10fdb5c272ab20eb527dc04997ce94518bb4ad35, deletion_comment = False)

333fred commented:
@AlekseyTs addressed feedback. I have an email out to the compat council for the breaking change.

---
In reply to: [742085337](https://github.com/dotnet/roslyn/pull/49804#issuecomment-742085337) [](ancestors = 742085337)

333fred commented:
Other PR is merged. I'll wait till review on this is done, then rebase before merging.

---
In reply to: [742084981](https://github.com/dotnet/roslyn/pull/49804#issuecomment-742084981) [](ancestors = 742084981)

AlekseyTs commented on src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs:
```diff
@@ -675,28 +659,109 @@ void writeChildrenProperty(AbstractNode type, List<Property> publicIOperationPro
+
+                        WriteLine($"case {slot - 1}:");
+                        Indent();
+                        WriteLine("return (false, int.MinValue, int.MinValue);");
```
>"return (false, int.MinValue, int.MinValue); [](start = 34, length = 44)

Is this going to cause a throw on a subsequent MoveNext call? This would violate the contract: "If MoveNext passes the end of the collection, the enumerator is positioned after the last element in the collection and MoveNext returns false. When the enumerator is at this position, subsequent calls to MoveNext also return false until Reset is called." I think once we reach "eof" slot, we should remain in this slot.

AlekseyTs replied:
I think something like ```$"return (false, {slot - 1}, {amount of items in the last property});"``` Alternatively, and perhaps simpler, for the case when the last property is an array, would be to "introduce" another case to handle "eof" - ``` $"case {slot}:"```. Both cases (```$"case {slot - 1}"``` and ```$"case {slot}:"``) would ```$"return (false, {slot}, 0);"```. Or ```$"case {slot - 1}"``` could simply $"goto case {slot};"


---
In reply to: [541126952](https://github.com/dotnet/roslyn/pull/49804#discussion_r541126952) [](ancestors = 541126952)

333fred replied:
I was thinking that this would be as simple as `return (false, previousSlot, previousIndex)`

---
In reply to: [541146522](https://github.com/dotnet/roslyn/pull/49804#discussion_r541146522) [](ancestors = 541146522,541126952)

AlekseyTs replied:
We want ```GetCurrent``` to throw on eof

---
In reply to: [541149260](https://github.com/dotnet/roslyn/pull/49804#discussion_r541149260) [](ancestors = 541149260,541146522,541126952)

AlekseyTs replied:
After looking at two distinct scenarios that we need to handle, I think an extra slot is needed in both cases in order for GetCurrent to work properly as well. And the code we need to generate is always the same:
```
case {slot - 1}:
    goto case {slot};
case {slot}:
    return (false, {slot}, 0);
default:
    throw ExceptionUtilities.UnexpectedValue((previousSlot, previousIndex));
``` 

---
In reply to: [541152163](https://github.com/dotnet/roslyn/pull/49804#discussion_r541152163) [](ancestors = 541152163,541149260,541146522,541126952)

AlekseyTs commented:
>             }

To clarify my comment in the class writer. I suggest to generate this here:
-Option 1:
```
                case 0:
                    return (false, 0, Operations.Length);
                default:
                    throw ExceptionUtilities.UnexpectedValue((previousSlot, previousIndex));
```

-Option 2:
```
                case 0:
                    goto case 1;
                case1:
                    return (false, 1, 0);
                default:
                    throw ExceptionUtilities.UnexpectedValue((previousSlot, previousIndex));
```

---
Refers to: src/Compilers/Core/Portable/Generated/Operations.Generated.cs:3352 in 0b096a1. [](commit_id = 0b096a10f54e1e2b7e32a3ffb825458c024712eb, deletion_comment = False)

AlekseyTs commented:
>             }

For this case, the last property is not an array, I suggest this code generation:
```
case 0:
    goto case 1;
case1:
    return (false, 1, 0);
default:
    throw ExceptionUtilities.UnexpectedValue((previousSlot, previousIndex));
```

---
Refers to: src/Compilers/Core/Portable/Generated/Operations.Generated.cs:3698 in 0b096a1. [](commit_id = 0b096a10f54e1e2b7e32a3ffb825458c024712eb, deletion_comment = False)

333fred commented:
>             }

Yep, this is about what I was thinking, expect that I didn't have a `goto case 1;`, I just let it fall through.

---
In reply to: [743363000](https://github.com/dotnet/roslyn/pull/49804#issuecomment-743363000) [](ancestors = 743363000)

---
Refers to: src/Compilers/Core/Portable/Generated/Operations.Generated.cs:3698 in 0b096a1. [](commit_id = 0b096a10f54e1e2b7e32a3ffb825458c024712eb, deletion_comment = False)

AlekseyTs commented:
>             }

> I just let it fall through.

That works too.

---
In reply to: [743366535](https://github.com/dotnet/roslyn/pull/49804#issuecomment-743366535) [](ancestors = 743366535,743363000)

---
Refers to: src/Compilers/Core/Portable/Generated/Operations.Generated.cs:3698 in 10fdb5c. [](commit_id = 10fdb5c272ab20eb527dc04997ce94518bb4ad35, deletion_comment = False)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/Operation.Enumerable.cs:
```diff
@@ -0,0 +1,116 @@
+        /// <summary>
+        /// Implements a struct-based enumerable for <see cref="Operation"/> nodes, using a slot-based system that tracks
+        /// the current slot, and the current index in the slot if the current slot is an immutable array. This type is not hardened
+        /// to default instances, and will null reference in these cases. Using <see cref="ChildOperations"/> and this is always
```
>to default instances [](start = 12, length = 20)

Default instances of what?

AlekseyTs commented on src/Compilers/Core/Portable/Operations/Operation.Enumerable.cs:
```diff
@@ -0,0 +1,116 @@
+        /// <summary>
+        /// Implements a struct-based enumerable for <see cref="Operation"/> nodes, using a slot-based system that tracks
+        /// the current slot, and the current index in the slot if the current slot is an immutable array. This type is not hardened
+        /// to default instances, and will null reference in these cases. Using <see cref="ChildOperations"/> and this is always
```
>Using [](start = 74, length = 5)

Is this a typo? I do not see ```ChildOperations``` used anywhere in this file.

333fred replied:
No, it's `Operation.ChildOperations`. I'll fully qualify it.

---
In reply to: [541170567](https://github.com/dotnet/roslyn/pull/49804#discussion_r541170567) [](ancestors = 541170567)

AlekseyTs replied:
> No, it's Operation.ChildOperations. I'll fully qualify it.

Still don't see this type **using** it.


---
In reply to: [541171643](https://github.com/dotnet/roslyn/pull/49804#discussion_r541171643) [](ancestors = 541171643,541170567)

333fred replied:
Moved to Operation.cs.

---
In reply to: [541173933](https://github.com/dotnet/roslyn/pull/49804#discussion_r541173933) [](ancestors = 541173933,541171643,541170567)

333fred commented:
>             }

Went with option 2.

---
In reply to: [743358865](https://github.com/dotnet/roslyn/pull/49804#issuecomment-743358865) [](ancestors = 743358865)

---
Refers to: src/Compilers/Core/Portable/Generated/Operations.Generated.cs:3352 in 0b096a1. [](commit_id = 0b096a10f54e1e2b7e32a3ffb825458c024712eb, deletion_comment = False)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/Operation.Enumerable.cs:
```diff
@@ -0,0 +1,116 @@
+                {
+                    case (false, _, _):
+                        _currentSlot = int.MinValue;
+                        _currentIndex = int.MinValue;
```
This doesn't look right, given the discussion around MoveNext implementation

333fred replied:
Yep, I've already corrected this locally.

---
In reply to: [541181111](https://github.com/dotnet/roslyn/pull/49804#discussion_r541181111) [](ancestors = 541181111)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationNodes.cs:
```diff
@@ -26,7 +23,27 @@ internal sealed class NoneOperation : Operation
+                0 when previousIndex + 1 < Children.Length
+                    => (true, 0, previousIndex + 1),
+                0 => (false, int.MinValue, int.MinValue),
+                _ => throw ExceptionUtilities.UnexpectedValue((previousSlot, previousIndex))
```
It looks like manual implementations have the same issue around not fulfilling the contact.

333fred replied:
Yep, also corrected locally.

---
In reply to: [541186396](https://github.com/dotnet/roslyn/pull/49804#discussion_r541186396) [](ancestors = 541186396)

AlekseyTs commented:
Done with review pass (commit 5)

AlekseyTs reviewed (approved):
LGTM (commit 7)

333fred commented:
@dotnet/roslyn-compiler for a second review.

chsienki reviewed (approved):
:shipit:
